Source/BoundaryConditions/WarpX_PEC.cpp

```diff
@@ -220,82 +230,105 @@ PEC::ApplyPECtoRhofield (amrex::MultiFab* rho, const int lev, PatchType patch_ty
         amrex::IntVect ref_ratio = ( (lev > 0) ? WarpX::RefRatio(lev-1) : amrex::IntVect(1) );
         domain_box.coarsen(ref_ratio);
     }
+    domain_box.convert(rho->ixType());
+
     amrex::IntVect domain_lo = domain_box.smallEnd();
     amrex::IntVect domain_hi = domain_box.bigEnd();
 
-    amrex::GpuArray<int, 3> fbndry_lo, fbndry_hi;
-    amrex::GpuArray<ParticleBoundaryType, 3> pbndry_lo, pbndry_hi;
-    for (int idim=0; idim < AMREX_SPACEDIM; ++idim) {
-        fbndry_lo[idim] = WarpX::field_boundary_lo[idim];
-        fbndry_hi[idim] = WarpX::field_boundary_hi[idim];
-        pbndry_lo[idim] = WarpX::particle_boundary_lo[idim];
-        pbndry_hi[idim] = WarpX::particle_boundary_hi[idim];
-    }
     amrex::IntVect rho_nodal = rho->ixType().toIntVect();
     amrex::IntVect ng_fieldgather = rho->nGrowVect();
 
+    // Create a copy of the domain which will be extended to include guard
+    // cells for boundaries that are NOT PEC
+    amrex::Box grown_domain_box = domain_box;
+
+    amrex::GpuArray<GpuArray<bool,2>, AMREX_SPACEDIM> is_pec;
+    amrex::GpuArray<GpuArray<amrex::Real,2>, AMREX_SPACEDIM> psign;
+    amrex::GpuArray<GpuArray<int,2>, AMREX_SPACEDIM> mirrorfac;
+    for (int idim=0; idim < AMREX_SPACEDIM; ++idim) {
+        is_pec[idim][0] = WarpX::field_boundary_lo[idim] == FieldBoundaryType::PEC;
+        is_pec[idim][1] = WarpX::field_boundary_hi[idim] == FieldBoundaryType::PEC;
+        if (!is_pec[idim][0]) grown_domain_box.growLo(idim, ng_fieldgather[idim]);
+        if (!is_pec[idim][1]) grown_domain_box.growHi(idim, ng_fieldgather[idim]);
+
+        psign[idim][0] = (WarpX::particle_boundary_lo[idim] == ParticleBoundaryType::Reflecting)
+                         ? 1._rt : -1._rt;
+        psign[idim][1] = (WarpX::particle_boundary_hi[idim] == ParticleBoundaryType::Reflecting)
+                         ? 1._rt : -1._rt;
+        mirrorfac[idim][0] = 2*domain_lo[idim] - (1 - rho_nodal[idim]);
+        mirrorfac[idim][1] = 2*domain_hi[idim] - (1 - rho_nodal[idim]);
```

Comment:
Should it be `2*domain_hi[idim] + (1 - rho_nodal[idim]);` (unless domain_box is converted to nodal like in ApplyPECtoJfield)?

Reply:
Yes, that is true. Thanks for catching it! I thought about the J case where I force the domain box to be nodal in order to get a consistent `mirrorfac` calculation for both the nodal and cell centered directions for the J-field and then just applied the same calculation here (where I just consider rho to be nodal since it usually is). 
```suggestion
        mirrorfac[idim][1] = 2*domain_hi[idim] + (1 - rho_nodal[idim]);
```